Controller panic on downscale

**What did you do to encounter the bug?**
I scaled my cluster from 5 to 3 replicas.

**What did you expect?**
A correct downscale, with a correct update of the automation configuration.

**What happened instead?**
The controller panics because of an out of range index reference and this log:
```
2022-02-10T17:10:44.215Z        INFO    manager/main.go:74      Watching namespace: sandbox-myservice-mongodb
I0210 17:10:45.265538       7 request.go:665] Waited for 1.03592925s due to client-side throttling, not priority and fairness, request: GET:https://10.32.0.1:443/apis/acme.cert-manager.io/v1beta1?timeout=32s
2022-02-10T17:10:45.820Z        INFO    manager/main.go:91      Registering Components.
2022-02-10T17:10:45.820Z        INFO    manager/main.go:104     Starting the Cmd.
2022-02-10T17:10:45.923Z        INFO    controllers/replica_set_controller.go:132       Reconciling MongoDB     {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:45.923Z        DEBUG   controllers/replica_set_controller.go:134       Validating MongoDB.Spec {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:45.923Z        DEBUG   controllers/replica_set_controller.go:143       Ensuring the service exists     {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.023Z        INFO    controllers/replica_set_controller.go:390       The service already exists... moving forward: services "myservice-mongodb5-svc" already exists {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.023Z        INFO    controllers/mongodb_tls.go:40   Ensuring TLS is correctly configured    {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.124Z        INFO    controllers/mongodb_tls.go:94   Successfully validated TLS config       {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.124Z        INFO    controllers/replica_set_controller.go:277       TLS is enabled, creating/updating TLS secret    {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.138Z        DEBUG   controllers/replica_set_controller.go:348       Enabling TLS on an existing deployment, the StatefulSet must be updated first   {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.138Z        INFO    controllers/replica_set_controller.go:288       Creating/Updating StatefulSet   {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.146Z        DEBUG   controllers/replica_set_controller.go:298       Ensuring StatefulSet is ready, with type: RollingUpdate {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.146Z        INFO    controllers/replica_set_controller.go:308       Creating/Updating AutomationConfig      {"ReplicaSet": "sandbox-myservice-mongodb/myservice-mongodb5"}
2022-02-10T17:10:46.173Z        DEBUG   scram/scram.go:165      Credentials have not changed, using credentials stored in: secret/myservice-mongodb5-root-user-scram-scram-credentials
2022-02-10T17:10:46.204Z        DEBUG   scram/scram.go:165      Credentials have not changed, using credentials stored in: secret/myservice-mongodb5-myservice-user-scram-scram-credentials
2022-02-10T17:10:46.228Z        DEBUG   scram/scram.go:165      Credentials have not changed, using credentials stored in: secret/myservice-mongodb5-prometheus-exporter-user-scram-scram-credentials
2022-02-10T17:10:46.229Z        DEBUG   controllers/replica_set_controller.go:429       AutomationConfigMembersThisReconciliation       {"mdb.AutomationConfigMembersThisReconciliation()": 4}
panic: runtime error: index out of range [3] with length 3

goroutine 320 [running]:
github.com/mongodb/mongodb-kubernetes-operator/pkg/automationconfig.(*Builder).Build(0xc0005dadf8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /workspace/pkg/automationconfig/automation_config_builder.go:251 +0x163f
github.com/mongodb/mongodb-kubernetes-operator/controllers.buildAutomationConfig(0x157bc9a, 0x10, 0xc00071e060, 0x1f, 0xc00024ccd8, 0x13, 0x0, 0x0, 0xc000047420, 0x1a, ...)
        /workspace/controllers/replica_set_controller.go:446 +0x8e5
github.com/mongodb/mongodb-kubernetes-operator/controllers.ReplicaSetReconciler.buildAutomationConfig(0x1b1eb70, 0xc0002a09e0, 0xc0001fa460, 0xc00000e0b0, 0xc0003ca0f8, 0x157bc9a, 0x10, 0xc00071e060, 0x1f, 0xc00024ccd8, ...)
        /workspace/controllers/replica_set_controller.go:520 +0x845
github.com/mongodb/mongodb-kubernetes-operator/controllers.ReplicaSetReconciler.ensureAutomationConfig(0x1b1eb70, 0xc0002a09e0, 0xc0001fa460, 0xc00000e0b0, 0xc0003ca0f8, 0x157bc9a, 0x10, 0xc00071e060, 0x1f, 0xc00024ccd8, ...)
        /workspace/controllers/replica_set_controller.go:413 +0x105
github.com/mongodb/mongodb-kubernetes-operator/controllers.(*ReplicaSetReconciler).deployAutomationConfig(0xc0005e7c88, 0x157bc9a, 0x10, 0xc00071e060, 0x1f, 0xc00024ccd8, 0x13, 0x0, 0x0, 0xc000047420, ...)
        /workspace/controllers/replica_set_controller.go:315 +0x336
github.com/mongodb/mongodb-kubernetes-operator/controllers.(*ReplicaSetReconciler).deployMongoDBReplicaSet.func1(0x1, 0x0, 0x0)
        /workspace/controllers/replica_set_controller.go:379 +0x85
github.com/mongodb/mongodb-kubernetes-operator/pkg/util/functions.runReversed(0xc0005e2150, 0x2, 0x2, 0x7, 0xc0003a6430, 0xc0000d2258)
        /workspace/pkg/util/functions/functions.go:29 +0x4c
github.com/mongodb/mongodb-kubernetes-operator/pkg/util/functions.RunSequentially(0xc0005e7c00, 0xc0005e2150, 0x2, 0x2, 0x1f, 0xc00024ccd8, 0x13)
        /workspace/pkg/util/functions/functions.go:11 +0x95
github.com/mongodb/mongodb-kubernetes-operator/controllers.(*ReplicaSetReconciler).deployMongoDBReplicaSet(0xc0005e7c88, 0x157bc9a, 0x10, 0xc00071e060, 0x1f, 0xc00024ccd8, 0x13, 0x0, 0x0, 0xc000047420, ...)
        /workspace/controllers/replica_set_controller.go:377 +0x135
github.com/mongodb/mongodb-kubernetes-operator/controllers.ReplicaSetReconciler.Reconcile(0x1b1eb70, 0xc0002a09e0, 0xc0001fa460, 0xc00000e0b0, 0xc0003ca0f8, 0x1b02178, 0xc000430090, 0xc000047420, 0x1a, 0xc00024ccd8, ...)
        /workspace/controllers/replica_set_controller.go:185 +0x12aa
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc00048a320, 0x1b02178, 0xc000430030, 0xc000047420, 0x1a, 0xc00024ccd8, 0x13, 0xc000430000, 0x0, 0x0, ...)
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:114 +0x247
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00048a320, 0x1b020d0, 0xc0000ae000, 0x16ea1e0, 0xc00049e8e0)
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:311 +0x305
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00048a320, 0x1b020d0, 0xc0000ae000, 0x0)
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2(0xc000594010, 0xc00048a320, 0x1b020d0, 0xc0000ae000)
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:227 +0x6b
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:223 +0x4d2
```

**Operator Information**
 - Operator version: 0.7.2
 - Operator image registry: Quay
 - MongoDB Image used: docker.io/mongo:5.0.4

**Kubernetes Cluster Information**
 - Distribution: none
 - Version:  v1.21.4

**Additional context**

At the moment of the log the `StatefulSet` has been updated to 4 replicas (from 5) and the automation config (as a `Secret`) has this configuration (5 Mongo ReplicaSet members):
```
kubectl get secret myservice-mongodb5-config -o=jsonpath='{.data.cluster-config\.json}' | base64 -d | jq '.replicaSets[0].members[]._id'
0
1
2
3
4
```

--- a/pkg/automationconfig/automation_config_builder.go
+++ b/pkg/automationconfig/automation_config_builder.go
@@ -283,7 +283,7 @@ func (b *Builder) Build() (AutomationConfig, error) {
 		processes[i] = *process
 
 		var horizon ReplicaSetHorizons
-		if b.replicaSetHorizons != nil {
+		if b.replicaSetHorizons != nil && i < len(b.replicaSetHorizons) {
 			horizon = b.replicaSetHorizons[i]
 		}
 

--- a/pkg/automationconfig/automation_config_test.go
+++ b/pkg/automationconfig/automation_config_test.go
@@ -159,6 +159,34 @@ func TestBuildAutomationConfigArbiters(t *testing.T) {
 	assert.Equal(t, 1, m[11].Votes)
 }
 
+func TestReplicaSetHorizonsScaleDown(t *testing.T) {
+	var expected ReplicaSetHorizons
+
+	horizons := []ReplicaSetHorizons{
+		{"horizon": "test-horizon-0"},
+		{"horizon": "test-horizon-1"},
+		{"horizon": "test-horizon-2"},
+	}
+	ac, err := NewBuilder().
+		SetName("my-rs").
+		SetDomain("my-ns.svc.cluster.local").
+		SetMongoDBVersion("4.2.0").
+		SetMembers(4).
+		SetReplicaSetHorizons(horizons).
+		Build()
+
+	assert.NoError(t, err)
+
+	for i, member := range ac.ReplicaSets[0].Members {
+		if i >= len(horizons) {
+			expected = nil
+		} else {
+			expected = ReplicaSetHorizons{"horizon": fmt.Sprintf("test-horizon-%d", i)}
+		}
+		assert.Equal(t, expected, member.Horizons)
+	}
+}
+
 func TestReplicaSetHorizons(t *testing.T) {
 	ac, err := NewBuilder().
 		SetName("my-rs").

--- a/pkg/automationconfig/automation_config_test.go
+++ b/pkg/automationconfig/automation_config_test.go
@@ -159,6 +159,46 @@ func TestBuildAutomationConfigArbiters(t *testing.T) {
 	assert.Equal(t, 1, m[11].Votes)
 }
 
+func TestReplicaSetMultipleHorizonsScaleDown(t *testing.T) {
+	var expected ReplicaSetHorizons
+
+	horizons := []ReplicaSetHorizons{
+		{
+			"internal": "test-horizon-0",
+			"external": "test-horizon-0",
+		},
+		{
+			"internal": "test-horizon-1",
+			"external": "test-horizon-1",
+		},
+		{
+			"internal": "test-horizon-2",
+			"external": "test-horizon-2",
+		},
+	}
+	ac, err := NewBuilder().
+		SetName("my-rs").
+		SetDomain("my-ns.svc.cluster.local").
+		SetMongoDBVersion("4.2.0").
+		SetMembers(4).
+		SetReplicaSetHorizons(horizons).
+		Build()
+
+	assert.NoError(t, err)
+
+	for i, member := range ac.ReplicaSets[0].Members {
+		if i >= len(horizons) {
+			expected = nil
+		} else {
+			expected = ReplicaSetHorizons{
+				"internal": fmt.Sprintf("test-horizon-%d", i),
+				"external": fmt.Sprintf("test-horizon-%d", i),
+			}
+		}
+		assert.Equal(t, expected, member.Horizons)
+	}
+}
+
 func TestReplicaSetHorizonsScaleDown(t *testing.T) {
 	var expected ReplicaSetHorizons
 

